### ISSUE TITLE
googler connection unexpectedly closed - ssl.SSLEOFError

### ISSUE THREAD

**Issue by sginsbourg** [Microsoft/WSL#3602]
~$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.1 LTS
Release:        18.04
Codename:       bionic

~$ googler --count 20 --exact --nocolor --noua --noprompt --debug "something" filetype:pdf
[DEBUG] googler version 3.5
[DEBUG] Python version 3.6.6
[DEBUG] User Agent is disabled
[DEBUG] Connecting to new host www.google.com
[DEBUG] setsockopt failed
Traceback (most recent call last):
  File "/usr/bin/googler", line 733, in new_connection
    self._conn.connect(self._notweak)
  File "/usr/bin/googler", line 268, in connect
    self.sock = ssl_context.wrap_socket(sock)
  File "/usr/lib/python3.6/ssl.py", line 407, in wrap_socket
    _context=self, _session=session)
  File "/usr/lib/python3.6/ssl.py", line 814, in __init__
    self.do_handshake()
  File "/usr/lib/python3.6/ssl.py", line 1068, in do_handshake
    self._sslobj.do_handshake()
  File "/usr/lib/python3.6/ssl.py", line 689, in do_handshake
    self._sslobj.do_handshake()
OSError: [Errno 0] Error

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/googler", line 2572, in <module>
    main()
  File "/usr/bin/googler", line 2546, in main
    repl = GooglerCmd(opts)
  File "/usr/bin/googler", line 1777, in __init__
    notweak=opts.notweak)
  File "/usr/bin/googler", line 676, in __init__
    self.new_connection(host, port=port, timeout=timeout)
  File "/usr/bin/googler", line 736, in new_connection
    raise GoogleConnectionError(msg)
__main__.GoogleConnectionError: Failed to connect to www.google.com: [Errno 0] Error.

**Comment by therealkenc** [comment]
Can't repro here on 18247.

**Comment by therealkenc** [comment]
It's a soft fail here (no `raise GoogleConnectionError(msg)` here with the CLI given), but the `setsockopt()` problem is:

```
212   connect(3, {sa_family=AF_INET, sin_port=htons(443), sin_addr=inet_addr("216.58.216.132")}, 16) = -1 EINPROGRESS (Operation now in progress)
212   clock_gettime(CLOCK_MONOTONIC, {tv_sec=178383, tv_nsec=686815800}) = 0
212   poll([{fd=3, events=POLLOUT|POLLERR}], 1, 45000) = 1 ([{fd=3, revents=POLLOUT}])
212   getsockopt(3, SOL_SOCKET, SO_ERROR, [0], [4]) = 0
212   setsockopt(3, SOL_TCP, TCP_DEFER_ACCEPT, [1], 4) = -1 ENOPROTOOPT (Protocol not available)
```
#1953

**Comment by sginsbourg** [comment]
Please advise on a work-around.
Thank you very much.

**Comment by therealkenc** [comment]
> Please advise on a work-around.

Change [this line](https://github.com/jarun/googler/blob/e704598f275f4e4bf8af100fe415ff6b8a981aaf/googler#L244) to "`if False:`" would probably do it. If that's the problem. Noting again that it isn't a hard fail here.

```
ken@DESKTOP-4UTIQSF:~$ googler --count 20 --exact --nocolor --noua --noprompt --debug "something" filetype:pdf
[DEBUG] googler version 3.5
[DEBUG] Python version 3.6.6
[DEBUG] User Agent is disabled
[DEBUG] Connecting to new host www.google.com
[DEBUG] setsockopt failed
[DEBUG] Fetching URL /search?ie=UTF-8&nfpr=1&num=20&oe=UTF-8&q=something+filetype%3Apdf
[DEBUG] Cookie: 1P_JAR=2018-10-08-10
[DEBUG] Response body written to '/tmp/googler-response-wg4ew6r5'.
 1 [PDF] something of for pdf - BBC
http://downloads.bbc.co.uk/worldservice/learningenglish/ask_about_english/pdfs/aae_sf_something_of_a.pdf
sentence we had to translate was the following: il était un peu ivrogne = he was something of a drunkard. If I've got
it right you need something of a + a noun…\
```

**Comment by sginsbourg** [comment]
Would you like to have a desktop sharing session for a repro ? Do let me know.

**Comment by therealkenc** [comment]
Windows version and an strace would be a start (you deleted the issue template). I take it a wget from a https URL works but googler does not?

**Comment by therealkenc** [comment]
Note to self #2846

**Comment by sginsbourg** [comment]
OS Name:                   Microsoft Windows 10 Home
OS Version:                10.0.17763 N/A Build 17763


$ strace -o googler.strace -f googler --debug
[DEBUG] googler version 3.5
[DEBUG] Python version 3.6.6
[DEBUG] Connecting to new host www.google.com
[DEBUG] setsockopt failed
Traceback (most recent call last):
  File "/usr/bin/googler", line 733, in new_connection
    self._conn.connect(self._notweak)
  File "/usr/bin/googler", line 268, in connect
    self.sock = ssl_context.wrap_socket(sock)
  File "/usr/lib/python3.6/ssl.py", line 407, in wrap_socket
    _context=self, _session=session)
  File "/usr/lib/python3.6/ssl.py", line 814, in __init__
    self.do_handshake()
  File "/usr/lib/python3.6/ssl.py", line 1068, in do_handshake
    self._sslobj.do_handshake()
  File "/usr/lib/python3.6/ssl.py", line 689, in do_handshake
    self._sslobj.do_handshake()
OSError: [Errno 0] Error

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/googler", line 2572, in <module>
    main()
  File "/usr/bin/googler", line 2546, in main
    repl = GooglerCmd(opts)
  File "/usr/bin/googler", line 1777, in __init__
    notweak=opts.notweak)
  File "/usr/bin/googler", line 676, in __init__
    self.new_connection(host, port=port, timeout=timeout)
  File "/usr/bin/googler", line 736, in new_connection
    raise GoogleConnectionError(msg)
__main__.GoogleConnectionError: Failed to connect to www.google.com: [Errno 0] Error.


$ wget www.google.com
--2018-10-09 08:13:30--  http://www.google.com/
Resolving www.google.com (www.google.com)... 216.58.212.132, 2a00:1450:400e:808::2004
Connecting to www.google.com (www.google.com)|216.58.212.132|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/html]
Saving to: ‘index.html.1’
index.html.1                      [ <=>                                              ]  13.36K  --.-KB/s    in 0s
2018-10-09 08:13:31 (57.0 MB/s) - ‘index.html.1’ saved [13684]


$ wget https://www.google.com
--2018-10-09 08:14:26--  https://www.google.com/
Resolving www.google.com (www.google.com)... 216.58.212.132, 2a00:1450:400e:808::2004
Connecting to www.google.com (www.google.com)|216.58.212.132|:443... connected.
Unable to establish SSL connection.

**Comment by therealkenc** [comment]
```
wget https://www.google.com
--2018-10-09 08:14:26-- https://www.google.com/
Resolving www.google.com (www.google.com)... 216.58.212.132, 2a00:1450:400e:808::2004
Connecting to www.google.com (www.google.com)|216.58.212.132|:443... connected.
Unable to establish SSL connection.
```

Aeight, you aren't able to establish a SSL connection period. You can try linking a [gist](https://gist.github.com/) of the output of `strace -f -o wget.strace wget https://www.google.com` just to have it on the books for completeness; though, in this case I don't think you are going see any syscall layer fails (noting fate #3130).

**Comment by sginsbourg** [comment]
$ strace -f -o wget.strace wget https://www.google.com --debug
DEBUG output created by Wget 1.19.4 on linux-gnu.

Reading HSTS entries from /home/shay/.wget-hsts
URI encoding = ‘UTF-8’
Converted file name 'index.html' (UTF-8) -> 'index.html' (UTF-8)
--2018-10-09 11:55:11--  https://www.google.com/
Resolving www.google.com (www.google.com)... 172.217.19.196, 2a00:1450:400e:808::2004
Caching www.google.com => 172.217.19.196 2a00:1450:400e:808::2004
Connecting to www.google.com (www.google.com)|172.217.19.196|:443... connected.
Created socket 3.
Releasing 0x00007fffdbb400b0 (new refcount 1).
Initiating SSL handshake.
SSL handshake failed.
Closed fd 3
Unable to establish SSL connection.

**Comment by sginsbourg** [comment]
https://gist.github.com/sginsbourg/9b8ded3760d54a7267c6cf943c1a0f4b

**Comment by therealkenc** [comment]
The **-o**utput is in the file `wget.strace`.

**Comment by sginsbourg** [comment]
wget.strace added to gist

**Comment by therealkenc** [comment]
Thanks. There are no WSL fails in there on a quick look; but that was expected. So you are down to (a) something not right in your userspace (ie Ubuntu), or (b) something external to WSL (VPN, firewall, 3rd party antivirus s/w, whatev). 

You can try installing 16.04 from the Store just as a fresh start start and do a `wget https://google.com`. Doing that has nothing to do with the version (`wget` works fine on all versions natch), it just gives you an unadulterated starting point without wiping your current install, and eliminates (a).

**Comment by sginsbourg** [comment]
Will try!
Thank you very much for your attention and good advice.

**Comment by sginsbourg** [comment]
BTW - 
On Windows CMD:

>wget https://www.google.com
--2018-10-09 22:01:26--  https://www.google.com/
Resolving www.google.com (www.google.com)... 172.217.19.196
Connecting to www.google.com (www.google.com)|172.217.19.196|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/html]
Saving to: 'index.html'

index.html                        [ <=>                                              ]  13.38K  --.-KB/s    in 0.05s

2018-10-09 22:01:30 (274 KB/s) - 'index.html' saved [13698]

**Comment by sginsbourg** [comment]
While on WSL:

$ wget --no-check-certificate https://www.google.com
--2018-10-09 22:04:00--  https://www.google.com/
Resolving www.google.com (www.google.com)... 172.217.19.196, 2a00:1450:400e:808::2004
Connecting to www.google.com (www.google.com)|172.217.19.196|:443... connected.
Unable to establish SSL connection.

**Comment by therealkenc** [comment]
> BTW -
> On Windows CMD:

Yeahbutt (a) there _is no_ `wget` in Windows™ CMD, and (b) wherever you got your win32 `wget` from (there are a number of sources), it has entirely different worldview of SSL/TLS than your Ubuntu userspace. SSL works on your Windows (naturally) or you wouldn't be able to read this message in Chrome or Edge from https://github.com (in your address bar).

We're treading dangerously out of WSL actionable territory here.

**Comment by sginsbourg** [comment]
Thanks again for all your help.

**Comment by therealkenc** [comment]
Just out of (very mild) curiosity.

```
$ powershell.exe "Get-CimInstance -Namespace root/SecurityCenter2 -ClassName AntivirusProduct" | grep displayName
```

At this point on this tracker we're left in need-repro territory. That's not on you: I know people experiencing this have no idea why (and neither do I). But there simply isn't enough information anywhere in this tracker to explain how to get from a clean PC with a clean WSL to a point where `wget` or `git` (or whatever) can't establish a SSL connection. Obviously that "works" or people's heads would explode. So the exercise remains finding the variable.

**Comment by andres-root** [comment]
The result for that command is:
displayName              : Windows Defender
displayName              : Kaspersky Total Security

**Comment by therealkenc** [comment]
> displayName : Kaspersky Total Security

Betcha.

**Comment by andres-root** [comment]
Actually you're right. I noticed that if I killed the AV process I was able to do a CURL. Then I found an option that filtered the HTTPS port. I disabled and it worked! Since I've been working on Linux environments for about 10 years now I didn't know that this kind of stuff could happen. Thanks so much for pointing that out. I posted this issue on Stack Overflow and Ask Ubuntu and nobody answered.

Thank you!
